### PR TITLE
[sw] Add Clang Device Software Build

### PR DESCRIPTION
**Pull request by lenary**
This PR adds a second device software build which uses clang (in addition to the GCC build). This build should be able to fail without causing an overall pipeline failure.

This is the next step of the toolchain migration.

(I've added some code to `mmio.h` which ensures that the build will fail on Clang, this should be reverted before this commit is merged).

### CODE REVIEW

File: ci/build-and-test-software.yml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    ./meson_init.sh -A ${{ parameters.extra_meson_init_args }}
+    ninja -C "$OBJ_DIR" all
+  displayName: Build embedded targets
+  continueOnError: ${{ eq(parameters.allowed_to_fail, 'true') }}
+- ${{ if eq(parameters.perform_tests, true) }}:
+  - bash: |
+      . util/build_consts.sh
+      ninja -C "$OBJ_DIR" test
+    displayName: 'Run unit tests'
+    continueOnError: ${{ eq(parameters.allowed_to_fail, 'true') }}
```
Comment (imphil) [inline ci/build-and-test-software.yml:33]: add trailing newline. (You really need to fix your editor config :-))

File: azure-pipelines.yml [hunk 2]
```diff
@@ -159,6 +160,18 @@ jobs:
...
+  strategy:
+    matrix:
+      gcc:
+        meson_init_args: ""
+        artifact_name: sw_build_gcc
+        allowed_to_fail: false
+      clang:
+        meson_init_args: "-t ${TOOLCHAIN_PATH}/meson-riscv32-unknown-elf-clang.txt"
+        artifact_name: sw_build_clang
+        allowed_to_fail: true
```
Comment (imphil) [inline azure-pipelines.yml:173]: All our Azure Pipelines variables are all-caps currently, let's do that here as well.

File: ci/build-and-test-software.yml [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+# These tasks build and test the OpenTitan software using `meson`.
+
+parameters:
+- name: extra_meson_init_args
+  displayName: "Extra Arguments for ./meson_init.sh"
+  type: string
+  default: ""
+- name: allowed_to_fail
+  displayName: "Is this build allowed to fail?"
+  type: string # This doesn't work as a boolean, of course
```
Comment (imphil) [inline ci/build-and-test-software.yml:14]: "of course"?

File: ci/build-and-test-software.yml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  default: 'false'
```
Comment (imphil) [inline ci/build-and-test-software.yml:15]: double quotes

File: ci/build-and-test-software.yml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+- name: perform_tests
+  displayName: "Should this build run tests?"
```
Comment (imphil) [inline ci/build-and-test-software.yml:17]: Please don't use quotes unless necessary (that's the style mcy used consistently throughout our pipelines config)

File: ci/build-and-test-software.yml [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    . util/build_consts.sh
+    ./meson_init.sh -A ${{ parameters.extra_meson_init_args }}
+    ninja -C "$OBJ_DIR" all
+  displayName: Build embedded targets
+  continueOnError: ${{ eq(parameters.allowed_to_fail, 'true') }}
+- ${{ if eq(parameters.perform_tests, true) }}:
+  - bash: |
+      . util/build_consts.sh
+      ninja -C "$OBJ_DIR" test
+    displayName: 'Run unit tests'
```
Comment (imphil) [inline ci/build-and-test-software.yml:32]:
```suggestion
    displayName: Run unit tests
```

**Comment by lenary** [comment]
I tried to use Azure Pipeline's matrix builds, but they're not polished enough that we can use them easily (because of all the variables that need to be propagated everywhere). I have given up on this and have just copy-pasted the GCC job for the moment. We will inevitably need to revisit this, but Azure doesn't really do the concept of "here's a parallel build that should be run, but should not stop our ability to merge".

As it is this minute, there is a compile error inserted into the code which ensures only the clang build fails. The status shows as "all checks have passed", which is the correct status for what we want. I am now going to remove the clang-only compile error from the software, and then this PR can be reviewed and merged.

File: azure-pipelines.yml [hunk 7]
```diff
@@ -182,6 +181,41 @@ jobs:
...
+  dependsOn: lint
+  condition: and(succeeded(), eq(dependencies.lint.outputs['DetermineBuildType.onlyDocChanges'], '0'))
+  pool:
+    vmImage: ubuntu-16.04
+  steps:
+  - template: ci/install-package-dependencies.yml
+  - bash: |
+      set -x
+      sudo util/get-toolchain.py \
+        --target-dir="${TOOLCHAIN_PATH}" \
```
Comment (imphil) [inline azure-pipelines.yml:199]: no need for curly braces here, but fine for consistency. If you feel like it, clean both targets up to remove them.

File: azure-pipelines.yml [hunk 8]
```diff
@@ -173,7 +173,6 @@ jobs:
       ./meson_init.sh -A
       ninja -C "$OBJ_DIR" all
     displayName: Build embedded targets
-  - template: ci/upload-artifacts-template.yml
```
[the remark this answers is not in the document]
Comment (lenary) [reply azure-pipelines.yml:176]: Done

**Comment by lenary** [comment]
Yep, this is good to go! :D

**Comment by imphil** [comment]
Clang!

**Pull request merged**